pydeseq2/dds.py

```diff
@@ -218,55 +239,190 @@ def __init__(
             raise ValueError(
                 "Either adata or both counts and metadata arguments must be provided."
             )
+        for col in self.obs.columns:
+            if col.strip() != col:
+                raise ValueError("Columns cannot contain leading or trailing spaces")
 
-        # Convert design_factors to list if a single string was provided.
-        self.design_factors = (
-            [design_factors] if isinstance(design_factors, str) else design_factors
+        self.design_factors = design_factors
+        self.continuous_factors = (
+            continuous_factors if continuous_factors is not None else []
         )
-        self.continuous_factors = continuous_factors
 
-        if self.obs[self.design_factors].isna().any().any():
-            raise ValueError("NaNs are not allowed in the design factors.")
-        self.obs[self.design_factors] = self.obs[self.design_factors].astype(str)
+        # If it's a list we convert it right back to a formulaic string
+        if isinstance(self.design_factors, list):
+            # Little bit of careful formating as below to remove space around : symbols
+            self.design_factors = [
+                re.sub(r"\ \:|\:\ ", ":", factor) for factor in self.design_factors
+            ]
```

Comment:
Is this worth the additional complexity? I would say it's ok to just support column names in `design_factors`. If someone wants interaction terms, they can just use a formula directly.

Reply:
I think it is but I am biased I didn't know formulaic before and I find it still a bit magical and unclear.

Reply:
I don't agree with this concern. Wilkinson formula are [pretty well defined](https://www.mathworks.com/help/stats/wilkinson-notation.html) and widely used, most prominently in R all over the place, but also in [matlab](https://www.mathworks.com/help/stats/fitlm.html#bunisqg-3), and Python e.g. [statsmodels](https://www.statsmodels.org/stable/examples/notebooks/generated/formulas.html). By supporting things like `"A:B"` as a design_factor, you basically endorse this notation already anyway. Arguably for everything more complex than just few covariates, the formula is more intuitive to read.

Reply:
You are right in the sense that it might be more intuitive and that they are well-defined however since they are very powerful they are very complex. One way this complexity manifests itself is there are subtle differences with the way R, formulaic and patsy (statsmodel) handle formulas due to the grammar being very powerful (see https://patsy.readthedocs.io/en/latest/R-comparison.html for instance). The way I see it is for now we would rather users not use them only those who want interaction terms and once we build the proper support for the whole grammar maybe deprecate the old way in the future.